apps/arweave/src/ar_nonce_limiter_client.erl

```diff
@@ -53,125 +60,114 @@ init([]) ->
 			gen_server:cast(?MODULE, pull)
 	end,
 	{ok, Config} = application:get_env(arweave, config),
-	Now =  erlang:system_time(millisecond),
-	RawPeersWithTimestamp = queue:from_list([{RawPeer, Now-?PULL_THROTTLE_MS}
-			|| RawPeer <- Config#config.nonce_limiter_server_trusted_peers]),
-	{ok, #state{ remote_servers = RawPeersWithTimestamp }}.
+	{ok, #state{ remote_servers = queue:from_list(Config#config.nonce_limiter_server_trusted_peers) }}.
 
 handle_call(Request, _From, State) ->
 	?LOG_WARNING([{event, unhandled_call}, {module, ?MODULE}, {request, Request}]),
 	{reply, ok, State}.
 
 handle_cast(pull, State) ->
-	#state{ remote_servers = Q } = State,
-	{{value, {RawPeer, Timestamp}}, Q2} = queue:out(Q),
-	Now = erlang:system_time(millisecond),
-	RotatedServers = queue:in({RawPeer, Now}, Q2),
-	case Now < Timestamp + ?PULL_THROTTLE_MS of
-		true ->
-			ar_util:cast_after(?PULL_THROTTLE_MS, ?MODULE, pull),
-			{noreply, State};
-		false ->
-			case ar_peers:resolve_and_cache_peer(RawPeer, vdf_server_peer) of
-				{error, _} ->
-					?LOG_WARNING([{event, failed_to_resolve_peer},
-							{raw_peer, io_lib:format("~p", [RawPeer])}]),
-					gen_server:cast(?MODULE, pull),
-					%% Push the peer to the back of the queue.
-					{noreply, State#state{ remote_servers = RotatedServers }};
-				{ok, Peer} ->
-					case ar_http_iface_client:get_vdf_update(Peer) of
-						{ok, Update} ->
-							#nonce_limiter_update{ session_key = SessionKey,
-									session = #vdf_session{
-											step_number = SessionStepNumber } } = Update,
-							State2 = update_latest_session_key(Peer, SessionKey, State),
-							UpdateResponse =
-								ar_nonce_limiter:apply_external_update(Update, Peer),
-
-							SessionFound = case UpdateResponse of
-								#nonce_limiter_update_response{ session_found = false } ->
-									false;
+	{RawPeer, State2} = get_raw_peer_and_update_remote_servers(State),
+	#state{ remote_servers = Q } = State2,
+	RotatedServers = rotate_servers(Q),
```

Comment:
This might double rotate the peers sometimes, right? e.g. if the timestamp has elapsed `get_raw_peer_and_update_remote_servers` will rotate, and then if there's an error querying RawPeer, it will rotate again. 

If they only have 2 remote peers, is it possible they could keep double rotate from the "bad" peer back to the "bad" peer?

what if `get_raw_peer_and_update_remote_servers` returns the `RotateServers` as well - i.e. if the timestap has elapsed it returns the new queue, and if the timestamp hasn't elapsed it return `rotate_servers(Q)`

Reply:
get_raw_peer_and_update_remote_servers rotates conditionally based on the elapsed time whereas on error we always rotate. That's why I use State2 returned by get_raw_peer_and_update_remote_servers in all success cases and RotatedServers - in the error cases. There is indeed a double-rotation bug here, however, because we need to rotate State#state.remote_servers, not State2#state.remote_servers.

Reply:
https://github.com/ArweaveTeam/arweave/pull/752/commits/6e297d9821ed81e2d147441ab0602ba82b98a762

Reply:
Looks good!